packages/material-ui/src/Paper/Paper.js

```diff
@@ -1,42 +1,71 @@
 import * as React from 'react';
 import PropTypes from 'prop-types';
 import clsx from 'clsx';
-import { useThemeVariants } from '@material-ui/styles';
-import withStyles from '../styles/withStyles';
+import { deepmerge } from '@material-ui/utils';
+import { unstable_composeClasses as composeClasses } from '@material-ui/unstyled';
+import experimentalStyled from '../styles/experimentalStyled';
+import useThemeProps from '../styles/useThemeProps';
 import useTheme from '../styles/useTheme';
+import { getPaperUtilityClass } from './paperClasses';
 
-export const styles = (theme) => {
-  const elevations = {};
-  theme.shadows.forEach((shadow, index) => {
-    elevations[`elevation${index}`] = {
-      boxShadow: shadow,
-    };
+const overridesResolver = (props, styles) => {
+  const { styleProps } = props;
+
+  return deepmerge(styles.root || {}, {
+    ...styles[styleProps.variant],
+    ...(!styleProps.square && styles.rounded),
+    ...(styleProps.variant === 'elevation' && styles[`elevation${styleProps.elevation}`]),
   });
+};
+
+const useUtilityClasses = (styleProps) => {
+  const { square, elevation, variant, classes } = styleProps;
 
+  const slots = {
+    root: [
+      'root',
+      variant,
+      !square && 'rounded',
+      variant === 'elevation' && `elevation${elevation}`,
+    ],
+  };
+
+  return composeClasses({ slots, classes, getUtilityClass: getPaperUtilityClass });
+};
+
+const PaperRoot = experimentalStyled(
+  'div',
+  {},
+  {
+    name: 'MuiPaper',
+    slot: 'Root',
+    overridesResolver,
+  },
+)(({ theme, styleProps }) => {
   return {
     /* Styles applied to the root element. */
-    root: {
-      backgroundColor: theme.palette.background.paper,
-      color: theme.palette.text.primary,
-      transition: theme.transitions.create('box-shadow'),
-    },
+    backgroundColor: theme.palette.background.paper,
+    color: theme.palette.text.primary,
+    transition: theme.transitions.create('box-shadow'),
     /* Styles applied to the root element unless `square={true}`. */
-    rounded: {
+    ...(!styleProps.square && {
       borderRadius: theme.shape.borderRadius,
-    },
+    }),
     /* Styles applied to the root element if `variant="outlined"`. */
-    outlined: {
+    ...(styleProps.variant === 'outlined' && {
       border: `1px solid ${theme.palette.divider}`,
-    },
+    }),
     /* Styles applied to the root element if `variant="elevation"`. */
```

Comment:
Keeping the comments in an interesting idea

Reply:
Don't want to delete those yet because of migration. Just to make sure it's all the same even if the styling system changed.